### PR TITLE
[CodeCompletion] Consolidate parameter list processing functions

### PR DESCRIPTION
**Pull request by rintaro**
There were 2 functions to output argument list (`addParameters` and `addParamPatternFromFunction `). Consolidate them (`addCallArgumentPatterns`) and consistently use it from every call like production (i.e. function call, constructor call, enum with associated values, subscript)

Also:
- Display mapped result type for `subscript`
   ```swift
   func foo(x: [Int]) {
      x #^COMPLETE^#
   }
   ```
   should display `Int` instead of `Element`.
- Don't suggest duplicated result for curried instance methods with defaulted parameters:
   ```swift
   class C {
       func foo(arg: Int, optArg: Int = 1) {}
   }
   C.#^COMPLETE^#
   ```
- Mitigate wrong argument label emission for curried instance methods:
   ```swift
   class C {
     func foo(x: Int) { }
     static func test() {
       self.foo(#^COMPLETE^#
     }
   }
   ```
   This used to end up with `foo(x: <#C#>)`

As side effects, this PR slightly changes parameter name handling. I think these are good for consistency.

- Instance method (`func methodName(x: Int) -> ...`) on meta type: `self` is now an inner name instead of (annotation)label.

    |  | old | new |
    |---|---|---|
    | sourcetext | `methodName(<#T##TheType#>)` | `methodName(<#T##self: TheType##TheType#>)` |
    | description | `methodName(self: TheType)` | `methodName(self: TheType)` (same) |

-  Now it shows inner names for subscript indices: `subscript(index: IndexType) -> ...` results: 

    |  | old | new |
    |---|---|---|
    | sourcetext | `[<#T##IndexType#>]` | `[<#T##index: IndexType##IndexType#>]` |
    | description | `[IndexType]` | `[index: IndexType]` |

### CODE REVIEW

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
@benlangmuir Could you review this?

File: lib/IDE/CodeCompletion.cpp [hunk 1]
```diff
@@ -2059,99 +2038,98 @@ class CompletionLookup final : public swift::VisibleDeclConsumer {
...
-        BodyParams = nullptr;
-
-        // Adjust to the "self" list if that is present, otherwise give up.
-        if (expectedNumParams == 1 && AFD->getImplicitSelfDecl())
-          SelfDecl = AFD->getImplicitSelfDecl();
-        BodyParams = nullptr;
-      }
-    }
+  /// Build argument patterns for calling. Returns \c true if any content was
+  /// added to \p Builder. If \p declParams is non-empty, tht size must match
```
Comment (benlangmuir) [inline lib/IDE/CodeCompletion.cpp]: Typo: tht

File: lib/IDE/CodeCompletionResultBuilder.h [hunk 2]
```diff
@@ -322,26 +322,20 @@ class CodeCompletionResultBuilder {
     addSimpleChunk(CodeCompletionString::Chunk::ChunkKind::CallParameterBegin);
 
     if (!Name.empty()) {
-      StringRef NameStr = Name.str();
-
-      // 'self' is a keyword, we cannot allow to insert it into the source
-      // buffer.
-      bool IsAnnotation = (NameStr == "self");
```
Comment (benlangmuir) [inline lib/IDE/CodeCompletionResultBuilder.h]: Wow, didn't know we had this hack here.  Glad you can simplify it away now.

**Review by benlangmuir — COMMENTED** [review]
Overall looks great.  I'd like to see a comment explaining `isImplicitlyCurriedInstanceMethod` though.

File: lib/IDE/CodeCompletion.cpp [hunk 3]
```diff
@@ -2263,10 +2239,45 @@ class CompletionLookup final : public swift::VisibleDeclConsumer {
...
+      // method or not.
+      auto getCurriedLevel = [](const AnyFunctionType *funcTy) -> unsigned {
+        unsigned level = 0;
+        while (auto result = funcTy->getResult()->getAs<AnyFunctionType>()) {
+          funcTy = result;
+          ++level;
+        }
+        return level;
+      };
+      bool isImplicitlyCurriedInstanceMethod =
```
[the remark this answers is not in the document]
Comment (rintaro) [reply lib/IDE/CodeCompletion.cpp]: Added explanation in a comment.
Only if we are calling instance method to meta type, the curried level matches.
```
struct S {
  static func foo() // interface type: (S.Type) -> () -> Int
  func bar()        // interface type: (S) -> () -> Int
}

S().foo // expression type: <<error type>>
S.foo   // expression type: () -> Int
S().bar // expression type: () -> Int
S.bar   // expression type: (S) -> () -> Int
```

> Also, is this check general? Can we replace the member function below with something like this, or is there some difference that prevents that?

I don't think so. The difference between `isImplicitlyCurriedInstanceMethod()` and here is that, in `isImplicitlyCurriedInstanceMethod(AbstructFuncDecl *)`, we know base expression type and the member decl, but here, we know the (not base) expression type and the decl.

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Review by benlangmuir — APPROVED** [review]

File: lib/IDE/CodeCompletion.cpp [hunk 4]
```diff
@@ -2268,10 +2244,49 @@ class CompletionLookup final : public swift::VisibleDeclConsumer {
...
+        !AFD->getInterfaceType()->is<AnyFunctionType>()) {
+      // Probably, calling closure type expression.
+      foundFunction(AFT);
+      addPattern();
+      return;
+    } else {
+      // Calling function or method.
+      foundFunction(AFD);
+
+      // FIXME: Hack because we don't know we are calling instance
```
Comment (slavapestov) [inline lib/IDE/CodeCompletion.cpp]: It should be easy to know this -- what is the underlying issue?

Comment (rintaro) [reply lib/IDE/CodeCompletion.cpp]: There's no issue. I will fix this when I have time.
I think it need to look into the original expression, right?

**Pull request merged**